### PR TITLE
refactor!: remove legacy Python CLI

### PR DESCRIPTION
**Pull request by willwashburn**
## Summary

- remove the legacy Python CLI, compatibility dispatcher, and Python regression suites
- replace the source-checkout dispatcher with a symlink-safe POSIX shell launcher for the Rust CLI
- simplify installs to expose only `ai-hist`; upgrades remove only recognized installer-managed legacy launchers and report every cleanup decision
- replace Python-based verification with isolated shell, SQLite, Node.js, and public CLI checks
- add a native CLI changelog and bump the pre-1.0 CLI version from 0.1.0 to 0.2.0

## Why this follows PR #70

PR #70 hardened JSONL cursor handling in the Python fallback because the fallback was still an officially installed compatibility path. This branch is based directly on the merge of PR #70, so that work remains in history. The delete/modify conflict is resolved here by intentionally ending fallback support now that every public command is implemented by the Rust CLI.

## Breaking changes

- `ai-hist-python` is removed
- the redundant public `ai-hist-rust` alias is removed
- `AI_HIST_CLI` is no longer supported and now fails with an explanatory message

The internal prebuilt asset remains `ai-hist-rust-bin`. Installer cleanup is content-and-size gated; unrecognized files are preserved with a warning.

## Review hardening

The follow-up addresses all actionable inline review findings: PATH-safe cloud setup, single-scheduler guidance, symlink resolution, explicit compatibility failure, safe legacy cleanup, environment-isolated installer tests, declared SQLite tooling, diagnostic E2E assertions, standalone dependency setup, robust WAL fixture readiness, and CLI-scoped release notes/versioning.

## Verification

- `cargo fmt --all -- --check`
- `cargo clippy --workspace --all-targets -- -A clippy::too_many_arguments -D warnings`
- `cargo test --workspace` (133 tests)
- `npm run lint && npm test` in `sdk-ts` (18 tests)
- `npm run build:debug` plus generated typings check in `crates/ai-hist-napi`
- `./scripts/verify-install.sh`
- `AI_HIST_RUST_BIN="$PWD/target/debug/ai-hist" ./scripts/verify-e2e.sh`

### CODE REVIEW

File: ai-hist [hunk 1]
```diff
@@ -1,161 +1,12 @@
...
-            str(RUST_MANIFEST),
-            "--",
-            *argv,
-        ]
-    return subprocess.run(cmd).returncode
-
-
-if __name__ == "__main__":
-    raise SystemExit(main())
+exec cargo run -q -p ai-hist-cli --manifest-path "$ROOT_DIR/Cargo.toml" -- "$@"
```
Comment (willwashburn) [inline ai-hist]: **`AI_HIST_CLI` becomes a silent no-op, and the launcher loses symlink resolution.**

Two regressions in this rewrite:

1. **Silent behavior switch.** `AI_HIST_CLI` is now ignored entirely. Anyone with `export AI_HIST_CLI=python` in their shell profile (the escape hatch this repo shipped and advertised in `sdk-ts/CHANGELOG.md` 0.4.x) gets silently switched to the Rust CLI with no message. The old dispatcher at least validated the value and errored on garbage. For a breaking removal, the launcher should say so for one release:

```sh
if [ -n "${AI_HIST_CLI:-}" ]; then
  echo "ai-hist: AI_HIST_CLI is no longer supported; the Python CLI was removed." >&2
  [ "$AI_HIST_CLI" = "python" ] && exit 2
fi
```

Failing loudly is the whole point of a breaking change — a user whose muscle memory says "set `AI_HIST_CLI=python` when the Rust path misbehaves" now has no signal that the lever is gone.

2. **`dirname "$0"` does not resolve symlinks.** The old dispatcher used `Path(__file__).resolve().parent`, so `ln -s ~/src/relayhistory/ai-hist ~/bin/ai-hist` worked. Now `ROOT_DIR` resolves to `~/bin`, and `--manifest-path ~/bin/Cargo.toml` fails with a confusing cargo error. If the dev launcher is only ever invoked in-tree, say so explicitly in the comment; otherwise resolve the link (a `while [ -L "$0" ]` loop, or `readlink -f` guarded for macOS).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: install.sh [hunk 2]
```diff
@@ -125,54 +125,18 @@ install_binary_launchers() {
...
-
-  warn "could not install ai-hist-python fallback from $url"
-  return 1
+remove_legacy_launchers() {
+  rm -f \
+    "$BIN_DIR/ai-hist-python" \
+    "$BIN_DIR/ai-hist-rust" \
+    "$INSTALL_DIR/ai-hist-python" \
+    "$INSTALL_DIR/ai-hist-wrapper"
 }
```
Comment (willwashburn) [inline install.sh:184]: **Unconditional `rm -f` of files the installer no longer owns, silently, and only in the prefix used *this* run.**

Three problems with this cleanup step:

1. **It deletes paths it cannot prove it created.** `$BIN_DIR/ai-hist-rust` is, as of this PR, a name the installer no longer manages. If a user hand-placed a real binary there (this repo's own `docs/cloud-sync.md` used to tell people to `cp target/release/ai-hist ~/.local/bin/…`), a routine `curl … | sh` upgrade silently deletes it. Gate the removal on content, e.g. only remove if the file is a small shim matching the generated `exec "${AI_HIST_RUST_BIN:-…}"` / `exec python3 …` pattern.

2. **It is silent.** Every other state-changing step in this script calls `info`/`warn`. Removing user-visible commands from `~/.local/bin` deserves at least `info "removed legacy launcher: $f"` — otherwise the only trace of a destructive action is in the diff of a script the user piped into `sh`.

3. **Cleanup is scoped to the current `BIN_DIR`/`INSTALL_DIR`.** The common upgrade path is fine, but anyone who originally installed under a different `AI_HIST_PREFIX` (or the default while now passing an explicit prefix) keeps a stale `$BIN_DIR/ai-hist-python` shim that `exec python3 "$INSTALL_DIR/ai-hist-python"` — a file this run just deleted if the prefixes happened to overlap, or that still exists and runs a CLI you no longer support. Either scan the default prefix as well, or print a migration note listing what to remove by hand.

Related, out of diff range: `resolve_source_dir` (L182) and the method dispatch (L226) still key "is this a source checkout?" on the presence of `$(script_dir)/ai-hist`. After this PR, `install_from_source` never reads that file — the detection now depends on a dev-only shell script that could be deleted without breaking anything else in the install path. Keying on `Cargo.toml` + `crates/ai-hist/` would be more honest.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/verify-install.sh [hunk 3]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+TMP="$(mktemp -d)"
+trap 'rm -rf "$TMP"' EXIT
```
Comment (willwashburn) [inline scripts/verify-install.sh:6]: **No environment isolation — an exported `AI_HIST_RUST_BIN` turns this entire script into a no-op verifier.**

The launcher this script installs is `exec "${AI_HIST_RUST_BIN:-$rust_bin}" "$@"`. So if the caller has `AI_HIST_RUST_BIN` exported, every assertion below (`--help`, `--version`, `sync`, `search`, and the `= "ai-hist 9.9.9"` check against the fake binary) runs the *caller's* binary and tells you nothing about what was actually installed. This is not hypothetical: the verification steps in this PR's own description export exactly that variable for `verify-e2e.sh`, and CI sets it on the e2e step. One reordering or one developer with it in their shell, and this script passes green while verifying nothing.

Same class of leak for `AI_HIST_DB` (line 40's `COUNT(*) = 2` is only meaningful against a fresh DB), `AI_HIST_PREFIX`, `AI_HIST_INSTALL_METHOD`, `AI_HIST_REF`, and `AI_HIST_NO_AUTOSYNC`.

```sh
unset AI_HIST_RUST_BIN AI_HIST_DB AI_HIST_PREFIX AI_HIST_BIN_DIR \
      AI_HIST_INSTALL_DIR AI_HIST_INSTALL_METHOD AI_HIST_BINARY_URL \
      AI_HIST_SOURCE_DIR AI_HIST_REF AI_HIST_VERSION AI_HIST_NO_AUTOSYNC
```

Also worth isolating `HOME` for the first install (lines 13–18) the way the second one does at line 67 — right now the source-install run executes against the real `$HOME`, which is the sort of thing that only bites once, in someone's dotfiles.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/verify-install.sh [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+AI_HIST_BIN_DIR="$BINARY_BIN" \
+AI_HIST_INSTALL_DIR="$BINARY_SHARE" \
+AI_HIST_NO_AUTOSYNC=1 \
+  sh "$ROOT/install.sh"
+
+test "$("$BINARY_BIN/ai-hist" --version)" = "ai-hist 9.9.9"
+test "$("$BINARY_BIN/ai-hist" recent)" = "[]"
+test ! -e "$BINARY_BIN/ai-hist-rust"
+test ! -e "$BINARY_BIN/ai-hist-python"
+test ! -e "$BINARY_SHARE/ai-hist-python"
```
Comment (willwashburn) [inline scripts/verify-install.sh]: **These three assertions are vacuous — they can never fail.**

The legacy files are only created in `SOURCE_BIN`/`SOURCE_SHARE` (lines 11–12). `BINARY_BIN` and `BINARY_SHARE` are fresh directories created by `install.sh` itself during this run, so `ai-hist-python` and `ai-hist-rust` were never there. Asserting their absence proves nothing about `remove_legacy_launchers` on the **prebuilt-binary upgrade path** — which is the path essentially every real user takes, and the one the PR description explicitly promises ("The installer removes those stale launchers during upgrades").

Delete the fake-legacy files into the binary dirs before line 66 so the assertion has teeth:

```sh
mkdir -p "$BINARY_BIN" "$BINARY_SHARE"
touch "$BINARY_BIN/ai-hist-python" "$BINARY_BIN/ai-hist-rust"
touch "$BINARY_SHARE/ai-hist-python" "$BINARY_SHARE/ai-hist-wrapper"
```

Two more gaps in the same vein:

- Line 12 creates `$SOURCE_SHARE/ai-hist-wrapper` but nothing asserts it was removed, even though `remove_legacy_launchers` targets it. Add `test ! -e "$SOURCE_SHARE/ai-hist-wrapper"`.
- Nothing tests the actual upgrade shape: an existing `$BIN_DIR/ai-hist` **shim from the old installer** (the one that branches on `AI_HIST_CLI=python`) being replaced. Writing that old shim by hand before the install and asserting the new content afterwards would cover the real migration.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -37,10 +33,8 @@ jobs:
       - name: Build Rust CLI for wrapper tests
         run: cargo build -q -p ai-hist-cli
 
-      - name: Python tests
-        run: |
-          python -m pip install pytest
-          AI_HIST_RUST_BIN="$PWD/target/debug/ai-hist" python -m pytest -q test_ai_hist.py test_cli_dispatch.py test_install.py
+      - name: Installer verification
+        run: ./scripts/verify-install.sh
```
Comment (willwashburn) [inline .github/workflows/ci.yml:45]: **Architectural: you removed a declared dependency (`setup-python`) and replaced it with an undeclared one (`sqlite3`).**

`setup-python@v5` was an explicit, pinned, cross-platform declaration of the toolchain the verification steps needed. What replaces it — `scripts/verify-install.sh` (line 40), `scripts/verify-e2e.sh` (lines 84, 107–120), and `sdk-ts/src/mcp-smoke.test.ts` — all now shell out to the `sqlite3` CLI, and **nothing in this workflow installs it**. It works today only because `ubuntu-latest` happens to ship it. When GitHub rolls the runner image, or a contributor runs `./scripts/verify-e2e.sh` on a minimal Linux box, the failure is `sqlite3: command not found` from inside a `$(…)` substitution — which under `set -e` surfaces as an assertion mismatch, not a missing-tool error.

Net: the dependency count didn't go down, it moved from declared to implicit. Either add `sudo apt-get install -y sqlite3` (and a `command -v sqlite3 || { …; exit 1; }` preflight in both scripts), or drop the CLI dependency altogether by doing the assertions through `ai-hist … --json` output, which is the interface you actually want to be testing anyway.

Two smaller CI-shape notes:

- `verify-install.sh:80` asserts `~/Library/LaunchAgents/com.ai-hist.sync.plist` is absent — a macOS-only path, asserted on `ubuntu-latest`. It can never fail here. The `test_install.py` case it replaces (`test_install_script_autosync_opt_out_leaves_launchd_untouched`) had the same limitation, so this isn't a regression, but it means `AI_HIST_NO_AUTOSYNC` has no real coverage; the Linux-side equivalent (systemd user unit / crontab entry) is what would actually be written here.
- The e2e step (line 54) runs with `AI_HIST_RUST_BIN` set, so the `cargo run` branch of the new `ai-hist` shell launcher — the default path for every developer in a source checkout — has zero CI coverage. One `./ai-hist --version` without that variable set would cover it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/verify-e2e.sh [hunk 6]
```diff
@@ -109,32 +104,24 @@ PY
...
-grok_session = conn.execute("SELECT cwd, git_branch, last_assistant_text FROM sessions WHERE source='grok' AND session_id='grok-e2e'").fetchone()
-if not grok_session or grok_session[0] != "/tmp/e2e/grok" or grok_session[1] != "main" or "grok assistant summary" not in (grok_session[2] or ""):
-    raise SystemExit(f"grok session metadata missing: {grok_session!r}")
-synthetic = conn.execute("SELECT COUNT(*) FROM history WHERE source='grok' AND prompt LIKE '%synthetic prompt%'").fetchone()[0]
-if synthetic:
-    raise SystemExit("grok synthetic prompt was imported")
-conn.close()
-PY
+sources=$(sqlite3 "$AI_HIST_DB" "SELECT group_concat(source, ',') FROM (SELECT DISTINCT source FROM history ORDER BY source)")
+test "$sources" = "claude,codex,cursor,grok,opencode,trajectory"
```
Comment (willwashburn) [inline scripts/verify-e2e.sh]: **`group_concat` ordering is explicitly arbitrary in SQLite, and the check silently changed from "subset" to "exact".**

Two independent problems in this two-line translation:

1. **Unreliable ordering.** SQLite's docs state the order of elements concatenated by `group_concat()` is arbitrary; an `ORDER BY` in the subquery does not constrain it (only the `ORDER BY` clause *inside* the aggregate, added in 3.44, does). This passes today by luck of the current query plan. Either use `sqlite3` in list mode and sort in the shell, or pin it properly:

```sh
sources=$(sqlite3 "$AI_HIST_DB" "SELECT DISTINCT source FROM history" | sort | paste -sd, -)
```

2. **Semantics flipped.** The Python it replaces computed `missing = expected - sources` — a *subset* check that failed only when an expected source was absent. This is exact equality, so the day someone adds a seventh source, this e2e goes red for a reason unrelated to their change, with no diagnostic (see my note on lines 113–117). If exact equality is intentional as a "no unexpected sources" guard, that's defensible — but it's a behavior change worth stating in the diff rather than smuggling in as a translation.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/verify-e2e.sh [hunk 7]
```diff
@@ -109,32 +104,24 @@ PY
...
+test "$codex_project" = "/tmp/e2e/codex"
+
+claude_session=$(sqlite3 "$AI_HIST_DB" "SELECT cwd || '|' || git_branch || '|' || last_assistant_text FROM sessions WHERE source='claude' AND session_id='claude-e2e'")
+test "$claude_session" = "/tmp/e2e/project|main|assistant summary"
+
+grok_session=$(sqlite3 "$AI_HIST_DB" "SELECT cwd || '|' || git_branch || '|' || last_assistant_text FROM sessions WHERE source='grok' AND session_id='grok-e2e'")
+test "$grok_session" = "/tmp/e2e/grok|main|grok assistant summary"
+
+synthetic=$(sqlite3 "$AI_HIST_DB" "SELECT COUNT(*) FROM history WHERE source='grok' AND prompt LIKE '%synthetic prompt%'")
+test "$synthetic" = "0"
```
Comment (willwashburn) [inline scripts/verify-e2e.sh]: **NULL-collapsing concatenation + zero diagnostics: these assertions can only tell you *that* something broke, never *what*.**

Three issues:

1. **`||` propagates NULL.** If `git_branch` (or `cwd`, or `last_assistant_text`) comes back NULL — precisely the metadata-backfill regression this assertion exists to catch — the whole expression is NULL and `sqlite3` prints an empty string. Same output as "row does not exist" and as "sync produced nothing at all". Three distinct failure modes collapse into one indistinguishable empty string. Use `-separator` with multiple columns, or `quote()`/`coalesce()` per column so a NULL is visible as `NULL` rather than erasing its neighbours.

2. **No failure message.** A bare `test A = B` under `set -e` exits silently; the CI log shows the script died and nothing else. The Python this replaces raised `SystemExit(f"claude session metadata missing: {claude_session!r}")` — it printed the actual value, which is what you need at 2am. A four-line helper restores it entirely:

```sh
assert_eq() { # label actual expected
  [ "$2" = "$3" ] || { echo "FAIL $1: got '$2', want '$3'" >&2; exit 1; }
}
```

3. **Substring → exact match (line 114/117).** The Python asserted `"assistant summary" not in claude_session[2]`, i.e. a *substring* check, deliberately tolerant of the CLI wrapping or truncating `last_assistant_text`. This now demands byte-exact equality. It passes against the current fixture, but it silently couples the e2e to an implementation detail the previous check went out of its way not to depend on. If exact matching is intended, fine — but it's a tightening, not a translation.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/verify-e2e.sh [hunk 8]
```diff
@@ -109,32 +104,24 @@ PY
...
 
 rust_ai_hist --db "$AI_HIST_DB" tag codex-e2e release-e2e --source codex
 rust_ai_hist --db "$AI_HIST_DB" search release --tag release-e2e --json
 
-(cd "$ROOT/sdk-ts" && npm ci && npm test)
+(cd "$ROOT/sdk-ts" && npm test)
```
Comment (willwashburn) [inline scripts/verify-e2e.sh]: **Dropping `npm ci` makes this script non-standalone — it now only works if something else already installed `sdk-ts/node_modules`.**

In CI this passes because the "TypeScript tests" step ran `npm ci` earlier in the same job. Run `./scripts/verify-e2e.sh` on a fresh clone — exactly what the PR description documents as the verification procedure — and `npm test` fails on a missing `vitest`/`node_modules`, in a script whose entire job is to be the end-to-end smoke test.

This change is unrelated to removing Python, so if the motivation was CI wall-clock, make it conditional rather than unconditional:

```sh
[ -d "$ROOT/sdk-ts/node_modules" ] || (cd "$ROOT/sdk-ts" && npm ci)
(cd "$ROOT/sdk-ts" && npm test)
```

Unrelated nit while you're in this file: `rust_ai_hist()` (lines 16–23) now duplicates the new `ai-hist` launcher's logic byte for byte, and the script uses both forms interchangeably (`"$ROOT/ai-hist" sync` on line 93 vs `rust_ai_hist …` on line 122). Since the launcher is now exactly `AI_HIST_RUST_BIN` → else `cargo run`, the helper can be deleted and every call site made `"$ROOT/ai-hist"` — which has the side benefit of putting the new launcher on the e2e path.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: sdk-ts/src/mcp-smoke.test.ts [hunk 9]
```diff
@@ -181,35 +181,20 @@ test('SDK fallback ingests OpenCode rows committed in WAL files', async () => {
...
-live = sqlite3.connect(db)
-assert live.execute("SELECT COUNT(*) FROM part").fetchone()[0] == 1
-live.close()
-ready.write_text("ready")
-time.sleep(60)
-`,
-    ],
-    { stdio: ['ignore', 'ignore', 'pipe'] },
-  );
+  const child = spawn('sqlite3', [dbPath], { stdio: ['pipe', 'ignore', 'pipe'] });
```
Comment (willwashburn) [inline sdk-ts/src/mcp-smoke.test.ts]: **Unguarded spawn of an undeclared binary, and no `-bail`, in the one test whose whole point is a race.**

Three concrete failure modes this introduces:

1. **No `error` listener on the child.** If `sqlite3` isn't on `PATH`, Node emits an `'error'` event on the ChildProcess with no listener attached → unhandled error, and the immediately following `child.stdin.write()` throws `EPIPE` on a destroyed stream. That surfaces as an opaque crash of the test runner rather than "sqlite3 is not installed". `python3` was at least guaranteed by `actions/setup-python`, which this PR also removes — so the new dependency is both undeclared in CI and unguarded here. Add `child.on('error', …)` and either skip or fail with a clear message.

2. **No `-bail`.** The `sqlite3` CLI reading a script from stdin prints errors to stderr and **keeps going**. If any `CREATE TABLE`/`INSERT` fails, execution still reaches `.shell touch` on line 196, the readiness file appears, and the test proceeds against an empty or partial DB. Pass `['-bail', dbPath]`.

3. **The dropped assertion is the important one.** The Python this replaces opened a *second* connection and asserted `SELECT COUNT(*) FROM part == 1` **before** signalling ready. That is what made "ready" mean "rows are committed and visible to another reader" — the exact property this WAL regression test exists to pin. Now "ready" only means "the shell got that far". Restore it in SQL before the `.shell` line:

```sql
SELECT CASE WHEN (SELECT COUNT(*) FROM part) = 1 THEN 'ok' ELSE raise_error END;
```
or more simply keep `-bail` and add `SELECT COUNT(*) FROM part;` with an explicit check, so a silent schema failure can't masquerade as a passing test.

Worth also noting `PRAGMA journal_mode=WAL;` writes its result to stdout, which is `'ignore'` — so a filesystem that silently refuses WAL (some tmpfs/overlay/network mounts) leaves this test green while no longer testing WAL at all. The Python version had the same blind spot, so not a regression, but since you're rewriting the setup it's cheap to assert the pragma's return value.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: sdk-ts/src/mcp-smoke.test.ts [hunk 10]
```diff
@@ -181,35 +181,20 @@ test('SDK fallback ingests OpenCode rows committed in WAL files', async () => {
...
+  child.stdin.write(`
+PRAGMA journal_mode=WAL;
+PRAGMA wal_autocheckpoint=0;
+CREATE TABLE session (id TEXT PRIMARY KEY, directory TEXT, time_created INTEGER);
+CREATE TABLE message (id TEXT PRIMARY KEY, session_id TEXT, time_created INTEGER, data TEXT);
+CREATE TABLE part (id TEXT PRIMARY KEY, message_id TEXT, session_id TEXT, time_created INTEGER, data TEXT);
+INSERT INTO session VALUES ('oc-ts-wal', '/tmp/opencode-ts', 1700000000000);
+INSERT INTO message VALUES ('msg-ts-wal', 'oc-ts-wal', 1700000001000, '{"role":"user"}');
+INSERT INTO part VALUES ('part-ts-wal', 'msg-ts-wal', 'oc-ts-wal', 1700000002000, '{"type":"text","text":"ts wal opencode prompt"}');
+.shell touch ${readyPath}
```
Comment (willwashburn) [inline sdk-ts/src/mcp-smoke.test.ts]: **Unquoted path interpolated into a shell command.**

`.shell touch ${readyPath}` passes the path through `sqlite3`'s dot-command to `/bin/sh` unquoted. `mkdtemp` under a `TMPDIR` containing a space (a custom `TMPDIR`, or a macOS user directory someone pointed it at) splits into two arguments, `touch` creates two wrong files, `readyPath` never appears, and the test hangs until `waitForFile` times out — with a timeout message that points at the SDK, not at the fixture.

Quote it: `.shell touch '${readyPath}'`.

Also relevant here: this relies on `.shell` being available at all. It's disabled whenever `sqlite3` runs in safe mode (`-safe`), and some distro/hardened builds restrict dot-commands that shell out. A file-based readiness signal that stays inside SQLite — writing a sentinel row and having the test poll for it, or `.output`/`.once` to the ready path — avoids depending on the CLI's ability to fork a shell.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: sdk-ts/CHANGELOG.md [hunk 11]
```diff
@@ -7,6 +7,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
+### Changed
+
+- Make the repository and installer Rust-only. The source-checkout launcher is
+  now a small POSIX shell script, and CI no longer requires a Python runtime.
+
+### Removed
+
+- Remove the legacy Python CLI, its compatibility dispatcher, and the redundant
+  public `ai-hist-python` / `ai-hist-rust` launchers. The installer cleans up
```
Comment (willwashburn) [inline sdk-ts/CHANGELOG.md]: **Architectural: a breaking change to the CLI is being recorded in the npm SDK's changelog, where the affected users will never see it.**

The audience for "the Python CLI is gone, `AI_HIST_CLI=python` no longer works, `ai-hist-python` and `ai-hist-rust` are deleted from your `~/.local/bin`" is people who installed via `curl … install.sh`. This file ships to npm consumers of the `ai-hist` package — a package whose own behavior is unchanged by this PR beyond comment edits. Meanwhile:

- there is no root or CLI-level `CHANGELOG.md`;
- `crates/ai-hist/Cargo.toml` stays at `0.1.0`, so `ai-hist --version` can't distinguish a pre- from a post-removal install — the version string is the one thing a user has to reason about a breaking migration, and it doesn't move;
- the PR is titled `chore:`, which for a removal of a shipped, documented user-facing entry point undersells it (release is `workflow_dispatch`-driven so nothing auto-derives from it, but the git history is the record either way);
- `README.md` is updated to describe the new launcher set, but nothing tells an existing user what happened to the two commands that disappeared from their `PATH`.

Concretely, I'd want at least: a version bump on the CLI crate, a `### BREAKING` heading rather than `### Removed` buried under `[Unreleased]`, and one line in the README's install section noting that upgrading removes `ai-hist-python`/`ai-hist-rust`. The installer deleting files from a user's `bin` directory is currently the *only* migration signal they get.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/verify-install.sh [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+{"text":"installer codex prompt","ts":1700000001,"session_id":"install-codex"}
+JSONL
+
+TEST_DB="$TMP/history.db"
+HOME="$TEST_HOME" AI_HIST_DB="$TEST_DB" OPENCODE_DB="$TMP/missing-opencode.db" \
+TRAJECTORY_ROOT="$TMP/trajectories" "$SOURCE_BIN/ai-hist" sync >/dev/null
+test "$(sqlite3 "$TEST_DB" 'SELECT COUNT(*) FROM history')" = "2"
+search_output=$(HOME="$TEST_HOME" AI_HIST_DB="$TEST_DB" "$SOURCE_BIN/ai-hist" search installer --json)
+grep -q 'installer claude prompt' <<<"$search_output"
+grep -q 'installer codex prompt' <<<"$search_output"
```
Comment (willwashburn) [inline scripts/verify-install.sh]: **You're asserting against SQLite internals through a CLI you don't ship, when the tool under test already emits JSON.**

Line 40 shells out to `sqlite3` (undeclared dependency — see my comment on `ci.yml`) to count rows, and lines 42–43 `grep` for substrings in a JSON document. Both weaken the test:

- `grep -q 'installer claude prompt' <<<"$search_output"` matches the string anywhere in the payload — in a `project` field, in an error message, in a field you didn't intend. It doesn't assert *how many* results came back or which source each belongs to.
- The row count via `sqlite3` couples an installer smoke test to the DB schema. If `history` is ever renamed or partitioned, this fails for a reason having nothing to do with the installer.

Both checks can go through the binary's own contract, which is also the surface you actually want covered:

```sh
count=$(… "$SOURCE_BIN/ai-hist" search installer --json | python3 -c 'import json,sys; print(len(json.load(sys.stdin)))')
```
— well, not `python3` anymore, obviously. `node -e` is already a CI dependency here, or `ai-hist stats --json` if it exposes a total. The point stands: the installer test should assert on the CLI's public output, not reach around it into SQLite with a third tool.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by willwashburn — COMMENTED** [review]
## Adversarial review

The direction is right and long overdue — one implementation beats two, and the dispatcher's `PYTHON_COMMANDS = set()` / `PYTHON_ONLY_FLAGS = {}` on `main` is proof the fallback was already dead weight. Deleting 6,566 lines to get there is the correct trade. My objection is not to the removal; it's that the **replacement verification is materially weaker than what it replaces**, in ways that are individually small and collectively mean this PR's safety net is largely decorative.

I've left 12 inline comments. The ones I'd block on:

- **`scripts/verify-install.sh:77-79`** — the three assertions covering legacy-launcher cleanup on the **prebuilt-binary path** can never fail. The fake legacy files are only created in the source-install dirs. The upgrade behavior the PR description promises to users is, in practice, untested on the path ~every user takes.
- **`scripts/verify-install.sh:2-6`** — no env isolation. Because the installed shim is `exec "${AI_HIST_RUST_BIN:-…}"`, an exported `AI_HIST_RUST_BIN` (which this PR's own verification steps and CI both set) makes every assertion in the file test the caller's binary instead of the install.
- **`sdk-ts/src/mcp-smoke.test.ts:184`** — the readiness signal no longer implies the rows are committed. The Python it replaces asserted `COUNT(*) == 1` through a second connection *before* signalling ready; without that (and without `-bail`), a failed schema setup still trips the ready flag and the WAL regression test stops testing the race it's named after.

### Architectural

**1. You traded a declared dependency for an undeclared one.** `setup-python@v5` was pinned and explicit. `sqlite3` — now required by `verify-install.sh`, `verify-e2e.sh`, and the Node WAL test — is installed by nothing and declared nowhere. It works because `ubuntu-latest` happens to ship it. The README's "Zero dependencies" claim survives for end users, but the contributor-facing dependency count went *up*, and the failures when it's absent are unreadable (`set -e` on an empty command substitution). Details on `ci.yml`.

**2. What is the parity evidence for the deleted `test_ai_hist.py`?** The other two suites die honestly — `test_cli_dispatch.py` tested a dispatcher that no longer exists, and `verify-install.sh` genuinely mirrors all three `test_install.py` cases. But `test_ai_hist.py` was 2,569 lines encoding *parser and sync semantics*: Grok synthetic-prompt filtering, Cursor project-path decoding lossiness, Codex seconds→ms conversion, session metadata backfill. Some of that is covered by the 133 Rust tests and some by `verify-e2e.sh`'s six-source fixture — but the PR asserts "every public command is implemented by the Rust CLI" (commands) where the risk is in the parsers (edge cases). A short mapping in the PR body — deleted Python test case → the Rust test that now owns it, and anything deliberately dropped — would make this reviewable. Right now it's taken on faith.

**3. This deletes an approach, not just a runtime — and that deserves stating.** PR #70 hardened byte-offset resume cursors: device/inode identity, prefix-hash validation, rotation/truncation detection, commit-time re-verification. That machinery was Python-only. The Rust path takes a different design entirely: full rescan of every source per sync, deduped by `UNIQUE(source, timestamp_ms, prompt)` + `INSERT OR IGNORE`. So this PR doesn't merely retire a redundant implementation; it retires incremental resume as a concept. Two questions that follow:

- The autosync service runs every 60s by default. Has full-rescan cost been measured against a large real history (a year of Claude/Codex JSONL, plus an OpenCode SQLite), rather than the six-row e2e fixture? Rescan-and-dedupe is a legitimately better design than resumable cursors *if* the constant factor holds at scale — and a worse one if it doesn't, with no fallback left.
- `UNIQUE(source, timestamp_ms, prompt)` with Codex timestamps at second granularity means two identical prompts in the same second from the same source collapse to one row. Pre-existing, but this PR makes it the only dedupe strategy in the product. Worth confirming it's intended rather than inherited.

**4. The migration story is thin for a breaking change.** No CLI version bump (`crates/ai-hist` stays `0.1.0`), the changelog entry lives in the npm SDK's file where installer users won't see it, `AI_HIST_CLI=python` becomes a silent no-op rather than a clear error, and the installer's cleanup only covers the prefix used on that run. Detail on `sdk-ts/CHANGELOG.md`, `install.sh`, and `ai-hist`.

**5. Scope.** `WS9_SCAFFOLD_PLAN.md` (425 lines) has nothing to do with removing Python. Bundling unrelated deletions into a breaking-change PR makes the diff harder to review and the revert harder to scope. Not worth re-rolling for, but worth noting.

### Credit where due

The `install_binary_launchers` unification across the source and prebuilt paths is a real simplification — the two paths previously generated shims with different indirection depths (`ai-hist` → `ai-hist-wrapper` → binary vs. `ai-hist` → binary), and collapsing them removes a whole class of stale-wrapper bug. The fake-`cargo`-on-`PATH` trick in `verify-install.sh` (proving the binary path never builds) is a nice touch, and the `sdk-ts` comment cleanup is careful — it correctly distinguishes "old DBs lack this index" (still true) from "Python wrote this DB" (no longer the framing), rather than blanket find-and-replace.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by willwashburn** [comment]
Addressed all actionable review findings in 46bb3ad. Compatibility and symlink behavior, installer cleanup safety, test isolation, declared SQLite tooling, E2E semantics and diagnostics, WAL fixture robustness, setup docs, and native CLI changelog/versioning are updated. Local gates are green: fmt, clippy, 133 Rust tests, 18 SDK/MCP tests, NAPI contract generation, installer verification, and full E2E.

**Comment by willwashburn** [comment]
Addressed all five findings from the latest Cubic review in bbfe11c: E2E now verifies the complete source set and exact session metadata, always performs a clean locked SDK install, installer cleanup reports preserved symlinks, and the WAL fixture uses a database-level readiness probe while keeping the writer open. Local installer, SDK/MCP (18 tests), shell syntax, and full E2E verification pass.

**Comment by willwashburn** [comment]
Addressed the latest Cubic and CodeRabbit findings in d4f171e. Legacy cleanup now requires all identifying markers for known Python/Rust launchers, regression tests preserve partial-marker user wrappers and verify the exact symlink target remains untouched, and E2E clears inherited Relaycast credentials before sync. Installer and full E2E verification pass.
